coding guidelines: 15.7: fix several cases of if else if constructs missing final else

jenmwms opened the pull request:
The following contained cases of if else if constructs that
were missing a final else. This commit adds else {} to
comply with coding guideline 15.7.
- arch/arm/aarch32/cortex_m
- include/logging and /sys/
- kernel/
- lib/os/
- subsys/

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
arch: arm: core: aarch32: cortex_m: debug: remove if...else if construct

z_arm_debug_monitor_event_error_check() was missing final
else statement in the if else if construct so violated guideline
15.7. This commit removes the else if for symmetry in the limited
early-exit conditions, rather than empty final else {}, to comply.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
arch: arm: core: aarch32: cortex_m: fault: fix if...else ifs

bus_fault() and hard_fault() were missing final else statement
in the if else if constructs. This commit adds non-empty else {}
to comply with coding guideline 15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
include: sys: time_units: fix missing final else

z_tmcvt() was missing final else statement in the
if else if construct. This commit removes the
else if in this small structure to comply with
guideline 15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
kernel: poll: fix missing final else

register_events() and signal_poll_event() missing final
else statement in the if else if construct. This commit adds
else {} to comply with coding guideline 15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
kernel: work: fix missing final else

work_queue_main() was missing final else statement
in the if else if construct. This commit adds else {}
to comply with coding guideline 15.7. Includes a
context-specific description of why this branch is empty.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
lib: os: various places fix missing final else

The lib/os/ had several places missing final else
statement in the if else if construct. This commit adds
else {} or simple refactor to comply with coding guideline 15.7.
- cbprintf_complete.c
- cbprintf_nano.c
- heap-validate.c
- heap.c
- onoff.c
- p4wq.c
- sem.c

Also resolves the checkpatch issue of comments should align * on
each line.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
testsuite: ztest: src: fix missing final else

cleanup_test() was missing final else statement in
the if else if construct. This commit adds else {}
to comply with coding guideline 15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
usb: usb_device: fix missing final else

set_endpoint() and reset_endpoint() were missing final
else statement in the if else if construct. This commit
adds a final else {} with assert to comply with coding
guideline 15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

ceolin reviewed (approved):
that is the kind of violation that would be nice to fix / enforce with coccinelle

enjiamai reviewed (commented):
Excuse me, I have a question about those changes. Is this necessary for passing the coding guideline? And do we have to adapt this to all of our code? Thank you!

pabigot commented on arch/arm/core/aarch32/cortex_m/debug.c:
```diff
@@ -42,6 +42,7 @@ bool z_arm_debug_monitor_event_error_check(void)
 		 * would mean the system would be stuck in an infinite loop).
 		 */
 		return true;
+	} else {
```
FWIW, MISRA says:

> A final else statement shall always be provided whenever an if statement is followed by a sequence of one or more else if constructs. **The else statement shall contain at least either one side effect or a comment.** *(emphasis added)*

so this doesn't actually address the violation in full.

jenmwms replied:
Thank you @pabigot ! I wondered about that. I'll fix it them to fully address the issue :)

pabigot reviewed (commented):
Just a nit.  I don't think these sorts of things improve quality, but that doesn't really matter.

jenmwms commented:
> that is the kind of violation that would be nice to fix / enforce with coccinelle

Indeed that is a great tool for this! I do not have it set up yet, and liked going through the source code. But I will work on coccinelle to be more efficient and avoid such simple mistakes. Thanks!

jenmwms commented:
> Excuse me, I have a question about those changes. Is this necessary for passing the coding guideline? And do we have to adapt this to all of our code? Thank you!

Hi @enjiamai - Yes, this is a Required rule in our guidelines. We are applying it in phases (https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html). For this PR, I'm helping to address some main areas like the kernel, arch, related includes and subsys. Hoping to expand my scope soon

andyross commented on kernel/timeout.c:
```diff
@@ -310,6 +310,7 @@ uint64_t z_timeout_end_calc(k_timeout_t timeout)
 		return UINT64_MAX;
 	} else if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
 		return z_tick_get();
+	} else {
```
If the previous if clauses were all early-exits like they are here, you can put the remainder of the block into the final else as a clearer way to conform to the rule.

andyross reviewed (approved):
It might not hurt to do some actual refactoring instead of adding an empty else clause.  A lot of these are early-exit tests (e.g. early returns, break statements), which means that the existence of the "else" part is mostly superfluous.  So if you have code that looks like:

```c
if (a) {
     return A();
} else if (b) {
    return B();
}
return C();
```

You can symmetrize it by doing something like:

```c
if (a) {
     return A();
} else if (b) {
    return B();
} else {
    return C();
}
```

Or alternately:

```c
if (a) {
     return A();
}
if (b) {
    return B();
}
return C();
```

jenmwms commented on include/sys/time_units.h:
```diff
@@ -98,6 +98,7 @@ static TIME_CONSTEXPR ALWAYS_INLINE uint64_t z_tmcvt(uint64_t t, uint32_t from_h
 		} else if (round_off) {
 			off = rdivisor / 2U;
+		} else {
 		}
```
@andyross I'm refactoring this one but want to check on the expected behavior. The argument descriptions say we can get "round_down" behavior if we set `round_up` and `round_off` to false. In that case, we do not enter either if condition here. Does that mean "round_down" is the default  (`off = 0`)? If so, we could assert in the final else {} or we could simplify to two equal-level if statements.

We should The argument descriptions above this say

```
round_up  - Return the ceiling of the resulting fraction
round_off - Return the nearest value to the resulting fraction
                (pass both round_up/off as false to get "round_down")

andyross replied:
This one is a little subtle.  That function you're looking at is actually intended as something more like a macro -- those arguments are (except for the time "t" and one crazy weirdness about runtime frequency and the hpet driver) all expected to be compile time constructs which the compiler will inline to eliminate most of the code paths when expanding all the specific conversion routines.

So by construction (you can actually see the perl code I used to emit all these things in a comment in that file), "round_up" and "round_off" will never be both true.  So the "else" here is superfluous.  It's perfectly equivalent to write:


```c
		if (round_up) {
			off = rdivisor - 1U;
		}
                if (round_off) {
			off = rdivisor / 2U;
		}
```

jenmwms commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -79,6 +79,7 @@ static int cleanup_test(struct unit_test *test)
 		ret = TC_FAIL;
+	} else {
 	}
 
```
This test could use some improvement. Perhaps I can do it here as part of refactoring approach to this patch for if...else if constructs. 

I do not understand why check for `!ret` like this with ret implementation. -- ! is a logical operator. !ret has value 0 if ret has value non-zero. The value is initialized as `int ret = TC_PASS` (non-zero). Would it be more informative to initialize as `ret=0`, then this check makes?

jenmwms replied:
Oh, `TC_PASS` expands to 0.

jenmwms commented on arch/arm/core/aarch32/cortex_m/debug.c:
```diff
@@ -37,7 +37,8 @@ bool z_arm_debug_monitor_event_error_check(void)
-	} else if (SCB->DFSR & SCB_DFSR_BKPT_Msk) {
+	}
+	if (SCB->DFSR & SCB_DFSR_BKPT_Msk) {
 		/* Treat BKPT events as an error as well (since they
```
Will this move away from else if slow it down too much?

andyross replied:
No, because the compiler won't execute the second test because the first ends in an early return anyway.

jenmwms commented on kernel/poll.c:
```diff
@@ -213,6 +213,8 @@ static inline int register_events(struct k_poll_event *events,
 			register_event(&events[ii], poller);
 			events_registered += 1;
+		} else {
+			__ASSERT("Unknown condition cannot register events.");
```
missing argument

jenmwms commented on subsys/usb/usb_device.c:
```diff
@@ -490,25 +491,26 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
+		if (ret_enable == -EALREADY) {
+			LOG_WRN("Endpoint 0x%02x already enabled", ep_cfg.ep_addr);
+		}
+			LOG_ERR("Failed to enable endpoint 0x%02x", ep_cfg.ep_addr);
```
fix indent

jfischer-no commented on subsys/usb/usb_device.c:
```diff
@@ -524,7 +526,7 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 {
 	struct usb_dc_ep_cfg_data ep_cfg;
-	int ret;
+	int ret = -EINPROGRESS;
```
Please no, this not necessary.

jfischer-no commented on subsys/usb/usb_device.c:
```diff
@@ -535,14 +537,15 @@ static bool reset_endpoint(const struct usb_ep_descriptor *ep_desc)
-	if (ret == -EALREADY) {
-		LOG_WRN("Endpoint 0x%02x already disabled", ep_cfg.ep_addr);
-	} else if (ret) {
+	if (ret < 0) {
```
No, this is wrong, do not change the logic here.

jfischer-no commented on subsys/usb/usb_device.c:
```diff
@@ -490,25 +491,26 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
-	if (ret == -EALREADY) {
-		LOG_WRN("Endpoint 0x%02x already configured", ep_cfg.ep_addr);
-	} else if (ret) {
+	ret_configure = usb_dc_ep_configure(&ep_cfg);
```
No, this is wrong, please do not change the logic here and below.

jenmwms commented:
Thanks for the feedback @andy and @jfischer-no ! I'll update this PR.

jenmwms committed:
kernel: timeout: fix missing final else

z_timeout_end_calc() was missing final else statement
in the if else if construct. This commit pulls the last
condition into a final else {} to comply with guideline
15.7.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

nashif commented on kernel/work.c:
```diff
@@ -646,6 +646,8 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
 			 */
 			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		} else {
+			__ASSERT(false, "Unknown condition in work queue.");
```
this is breaking tests, so obviously we there is a case where the above conditions are not met which seems to be correct.

jenmwms replied:
Thank you! I will fix this.

jenmwms commented on kernel/poll.c:
```diff
@@ -213,6 +213,8 @@ static inline int register_events(struct k_poll_event *events,
 			register_event(&events[ii], poller);
 			events_registered += 1;
+		} else {
+			__ASSERT(false, "Unknown condition cannot register events.");
```
Looks like this is problematic and not appropriate assertion here, similar to work.c. I'll fix it here and below.

pabigot commented on kernel/work.c:
```diff
@@ -646,6 +646,8 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
 			 * submissions.
 			 */
 			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		} else {
```
The top commit at https://github.com/pabigot/zephyr/commits/pr/33198 provides a more helpful description, if you'd like to incorporate it.

jenmwms replied:
Thank you! I will cherry-pick to include this improvement :)

jenmwms commented:
> ```c
> if (a) {
>      return A();
> } else if (b) {
>     return B();
> } else {
>     return C();
> }
> ```

@andyross - this option may not work in the situation where the entire contents of the function are contained in the if... else if... else structure. Do you think we would get a 'missing return' error? I ask because the function is defined to have a return, but code scan tools might think that there is a path which might not yield a return for the function, so calls it an error.

jenmwms commented:
Removed the commit number from the subject line of the fixup commit. It caused compliance check error.

ioannisg commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -675,7 +676,8 @@ static uint32_t hard_fault(z_arch_esf_t *esf, bool *recoverable)
 		PR_EXC("  Bus fault on vector table read");
-	} else if ((SCB->HFSR & SCB_HFSR_FORCED_Msk) != 0) {
+	}
+	if ((SCB->HFSR & SCB_HFSR_FORCED_Msk) != 0) {
```
This does change the behavior, slightly, right? Cause originally, if the fault is on Vector Table read, the HF escalation conditional is not processed.

jenmwms replied:
Thank you! I misunderstood this slightly. I will revert the change to preserve the logic. Looks like I need to go with a final non-empty else{} (see other comment for our resolution on contents).

ioannisg replied:
Thanks for the fix.

pabigot commented on kernel/work.c:
```diff
@@ -647,7 +647,10 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
 			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
 		} else {
-			; /* Final statement (see coding guideline 15.7) */
+			/* No work is available and no queue state requires
```
Please squash this commit with the earlier one where the comment replaced here was introduced.

jenmwms replied:
Ok, I can do that. Is it to help keep the commits cleaner? jw. Thank you!

pabigot replied:
Yes, it's better not to leave false-starts in the history within a single PR.

jenmwms commented:
I cannot reproduce the doc and build errors reported by CI, and they seem unrelated. Maybe we can trigger it again?

nashif commented:
try to rebase against latest master

dcpleung commented:
tried latest master and doc build ran fine without errors.

jenmwms commented:
Rebased. Thank you!

enjiamai commented on lib/os/sem.c:
```diff
@@ -74,8 +74,9 @@ int sys_sem_give(struct sys_sem *sem)
 	} else if (old_value >= sem->limit) {
 		return -EAGAIN;
+	} else {
+		;
```
Excuse me, @jenmwms, I think if we add an "else block" without doing things here, the initial value of variable ret (in line 64) should not be zero, it should be -1 or some error value.  How do you think?

jenmwms replied:
I don't understand the issue with this patch. This final else does not change the logic. Are you suggesting to change the initial value of ret? That sounds like a potential bug and out of scope for this patch.

enjiamai replied:
Yes, @jenmwms  I agree with you. That seems like a bug even without this patch. I think it needs another fix for this.

enjiamai reviewed (approved):
no other questions, thanks.

nashif commented:
I need an ack on the changes to `lib/os/cbprintf_complete.c` from @nordic-krch
